ford/sourceform.py

```diff
@@ -1055,39 +1097,92 @@ def filter_public(collection: dict) -> dict:
                 typelist[dtype] = set([])
         typeorder = toposort.toposort_flatten(typelist)
 
+        # Correlate types
+        for dtype in typeorder:
+            if dtype in self.types:
+                dtype.correlate(project)
+
         # Match up called procedures
         if hasattr(self, "calls"):
             tmplst = []
             for call in self.calls:
-                call_low = call.lower()
+                call_low = call["name"].lower()
                 argname = False
                 for a in getattr(self, "args", []):
                     # Consider allowing procedures passed as arguments to be included in callgraphs
                     argname |= call_low == a.name.lower()
                 if hasattr(self, "retvar"):
                     argname |= call_low == self.retvar.name.lower()
+                
+                # travers the call chain of the call to descover the context the call is made on
+                context = self
+                call_chain = call["call_chain"]
+                if len(call_chain) > 0:
+                    class NoLabel(Exception):
+                        pass
+                    try:
+                        call_type = None
+                        # strip off the "type()" or "class()" if it's there
+                        strip_type = lambda s: re.match(r"^(type|class)\((.*?)(?:\(.*\))?\)$", s, re.IGNORECASE).group(2) if re.match(r"^(type|class)\((.*?)(?:\(.*\))?\)$", s, re.IGNORECASE) else s
+                        try: # try call type is a variable
+                            vars = self.all_vars
+                            if hasattr(self, "args"):
+                                vars = {**vars, **{a.name.lower(): a for a in self.args}}
+                            if hasattr(self, "retvar"):
+                                vars = {**vars, **{self.retvar.name.lower(): self.retvar}}
+                            call_type = vars[call_chain[0].lower()].full_type
+                            call_type = strip_type(call_type)
+                            call_type = self.all_types[call_type.lower()]
```

Comment:
There's quite a lot of calls to `str.lower()` -- I'd add that call to `strip_type`, and see if there's any intermediate variables that can be `lower`ed to save calling it multiple times on the same variable (e.g. `call_chain[0].lower()` below)